### PR TITLE
feat(cli): wire the CLI up to the Pipecat Context Hub

### PR DESCRIPTION
**Pull request by markbackman**
`pipecat init` writes `AGENTS.md` telling your coding agent to query the Pipecat Context Hub — and then the CLI didn't ship it. This closes that: the hub comes with `pipecat-ai[cli]`, `init` registers it with the agents you have, and the CLI tells you when the local index has drifted from the project you're in.

```console
$ pipecat init .
✔ Wrote AGENTS.md
✔ Wrote CLAUDE.md
? How do you want to build? Build with a coding agent (recommended)
✔ Wrote GETTING_STARTED.md
✓ Registered the Context Hub MCP server

  Your agent uses up-to-date knowledge of Pipecat APIs from Context Hub's
  local index, rather than relying on stale training data.
? Build it now? (3+ minutes, ~900 MB of disk) (Y/n)

⠹ Building the Context Hub index... · Source ingest (pipecat-ai/pipecat): upserted=6899
✓ Built the Context Hub index (2m 46s)
```

## Shipping the hub with the CLI

`pipecat-ai-context-hub` joins the `cli` extra. The hub's ONNX migration removed `torch`, which is what makes this affordable:

| | Size | Packages |
|---|---|---|
| `pipecat-ai[cli]` | 441 MB | 137 |
| `+ pipecat-ai-context-hub` | 636 MB | 255 |
| **marginal** | **+195 MB** | |

Measured on macOS arm64. **Bot runtimes are unaffected** — `cli` is optional precisely so the runtime and base-image install stay lean. `evals` inherits it, which is accepted rather than splitting `cli` into lean and full variants: that would reintroduce conditional installation to save 195 MB for users already pulling two local speech-model stacks.

`context-hub` also leaves `_KNOWN_EXTENSIONS`. That registry advertises plugins you must install separately; the hub is no longer one, and its stub would claim an install step that doesn't exist. It still appears in `pipecat --help` because it genuinely mounts via its entry point.

## Setup from `init`

Two actions, only one of them a question.

**Registration** runs on every qualifying invocation — instant, idempotent, and necessary per project, since Claude Code's MCP config is project-scoped.

**The index** is the only prompt, and only when none exists. An index can't exist unless someone ran `refresh`, so its presence *is* the "already dealt with" flag — and it's user-level, so the question doesn't return in other projects. It defaults to yes because `serve` refuses to start against an empty index, making "registered but not indexed" the worst reachable state.

| Path | Hub setup |
|---|---|
| `init` → "Build with a coding agent" | register + offer the index |
| `init .` non-interactive, no scaffold flags | register, print the index command, never block |
| `init . --transport …` (scaffold from flags) | **none** — the generated README is the start-here |
| `init` → "Scaffold a runnable bot now" | **none** — same reason |
| `init quickstart` | **none** — a fixed preset must stay a short path to a running bot |

`--no-context-hub` opts out anywhere.

Scaffolding from flags is the invocation `AGENTS.md` tells coding agents to use, so the guide now covers what an agent should do about it: if the session has no Context Hub tools, register the server so later sessions have it, *offer* a restart rather than pressing for one, and use the CLI meanwhile. The shell commands run the same handlers and return byte-identical JSON, so a session without the MCP tools isn't degraded — a restart is a convenience, not a fix.

Registration reports what came of it. Cursor, VS Code, and Zed are configured by hand rather than through a CLI, so `init` points at `pipecat context-hub install` to print the config to paste; a client CLI that rejects the registration reports why rather than passing in silence.

The build shows the hub's newest log line beside a spinner instead of its full stream — enough to see which source is indexing during the stretches of tens of seconds when nothing else surfaces. Deliberately no ETA: most of the time goes on embedding ~17k records on the CPU, so the total tracks single-core speed and a fixed estimate would read as an overrun on a slower machine. A failed build surfaces the hub's output and how to retry.

## The freshness notice

A stale index is invisible. The agent answers confidently either way — you find out when the generated code doesn't compile against the API you actually have.

```
⚠ Pipecat Context Hub index is 42 days old — run `pipecat context-hub refresh` so coding agents don't cite stale APIs.
⚠ Pipecat Context Hub index was built for pipecat-ai 1.6.0, but this project uses 1.8.0 — run `pipecat context-hub refresh` so coding agents match your version.
```

### Why the version comparison is so conservative

A false alarm trains people to ignore the notice, which costs more than the notice is worth. So it:

- compares **`major.minor` only** — `1.6.1.dev55` against `1.6.0` is silent, the overwhelmingly common case
- stays silent on an **editable pipecat checkout** — the developer *is* pipecat, and no released version describes their tree
- consults only `$VIRTUAL_ENV` and `./.venv`, never walking up parents, which would find an unrelated venv in a monorepo and warn about a project you aren't working on
- allows **a minor of slack when the index tracks the default branch**. The recorded tag is a floor, not an identity — an index 55 commits past `v1.6.0` may already contain 1.7.0's code. Without slack this fires at every developer working from a source checkout
- warns **only when the project is ahead**. An index built from `main` legitimately runs ahead of released users; a superset index isn't a correctness problem worth interrupting anyone over
- is silent on anything unparseable

Deliberately *not* `importlib.metadata.version("pipecat-ai")`: the CLI is normally a global `uv tool`, so that reports the tool's own version rather than the project's.

### How the check stays cheap

It reads the hub's `index_metadata` table directly, read-only, with stdlib `sqlite3` — no new dependency, sub-millisecond.

Not via the hub's own `status` command: every in-process hub query constructs an `IndexStore`, which opens ChromaDB even for lookups needing no embeddings. That would cost more than most commands being run. The hub publishes that table as a [documented read contract](https://github.com/pipecat-ai/pipecat-context-hub/pull/101) precisely so consumers like this one don't have to. The database is WAL, so a read-only open with `timeout=0` neither blocks nor is blocked by a concurrent refresh or a running MCP server, and sees only committed state.

It lives in `run()` rather than the group callback: click answers a bare `pipecat --help` eagerly and never invokes the callback, so a check placed there silently misses it.

### Noise control

- **Silent when no index exists**, so users who don't run the hub never see it.
- stderr only, so stdout stays parseable.
- `PIPECAT_HUB_CHECK=0` switches it off.
- The staleness threshold reads the hub's own `PIPECAT_HUB_STALE_AFTER_DAYS` (default 7), so tuning it moves both surfaces together instead of leaving them to disagree.
- Any failure — corrupt database, locked file, odd layout, unexpected exception — degrades to silence. A hint must never break the command you asked for.

## Verification

`ruff check` and `ruff format --check` clean; **379 passed** in `tests/cli/`.

Freshness tests build a real SQLite index rather than mocking the reader, so they exercise the same contract an installed hub publishes: missing index, corrupt database, missing table, contract version newer than understood, staleness above/below threshold, `nan`/`inf`/garbage thresholds, the disable switch, dist-info parsing, editable checkouts, Windows venv layout, `$VIRTUAL_ENV` precedence, and each version-comparison rule above.

Beyond the suite, each `init` path was exercised end to end against the published hub 0.5.1 — with an agent CLI present, with none installed, with one that rejects the registration, with and without an existing index, plus `quickstart` and `--no-context-hub`. The index build was run cold (17,359 records, 922 MB) and the interactive prompt and spinner driven through a pty.

Packaging was checked separately, in an isolated `UV_TOOL_DIR`, since everything else runs through the dev venv:

```console
$ uv tool install "/path/to/pipecat[cli]"
Installed 2 executables: pc, pipecat

$ pipecat context-hub install --print-config | head -1
Server command: /…/pipecat-ai/bin/python -m pipecat_context_hub serve
```

The extra resolves `pipecat-ai-context-hub 0.5.1`, the plugin mounts, and the registered command is absolute. Note the exposed executables: `pipecat-context-hub` is **not** among them, because `uv tool` exposes only the host package's scripts — which is exactly why the server command has to name an interpreter rather than a console script.

## Requires hub 0.5.1

Registration outcomes are read from `install`'s exit code, which [pipecat-ai/pipecat-context-hub#109](https://github.com/pipecat-ai/pipecat-context-hub/pull/109) adds along with two fixes this depends on: the registered server command is pinned to an absolute interpreter rather than a bare console-script name, and reinstalling repairs a stale registration instead of skipping it. 0.5.1 is released and the floor is pinned here, so this is ready to land on its own.

## Not verified

- Pasting the printed config into Cursor, VS Code, or Zed and watching the server start. The content is checked, and `install` exits `3` to say it printed rather than configured; the round trip itself isn't.
- Marginal install size on Linux — the +195 MB above is macOS arm64, and the hub's own figures differ by platform.
- What `uv sync --all-extras` now costs a contributor environment and CI.

### CODE REVIEW

File: src/pipecat/cli/agent_templates/GETTING_STARTED.md
Comment (kompfner) [inline src/pipecat/cli/agent_templates/GETTING_STARTED.md]: What if we made installing context hub just part of `pc init`? Then we wouldn't necessarily need this section at all.

There's a number of things to consider here. Some things we discussed:

- if coding agent is what's calling `pc init` on the user's behalf, we need to present a message to the user like "restart your coding agent now to use the Pipecat Context Hub"
- `pc init` without having installed with `--with pipecat-ai-context-hub` should show user a prompt "we noticed you don't have context hub installed; install again with `--with pipecat-ai-context-hub` (recommended) or proceed without it?"
- context hub doesn't support 3.15...

File: src/pipecat/cli/agent_templates/GETTING_STARTED.md [hunk 1]
```diff
@@ -11,18 +11,20 @@ confidently writing outdated APIs from memory. Set up the **Pipecat Context
...
 
 ```bash
-# One-time index build, using latest package; allow a few minutes
-uvx pipecat-ai-context-hub@latest refresh
-# Add the MCP server (use the line for your agent)
-claude mcp add pipecat-context-hub -- uvx pipecat-ai-context-hub serve   # Claude Code
-codex mcp add pipecat-context-hub -- uvx pipecat-ai-context-hub serve    # Codex
+# Co-install the hub with the CLI, then let it do the setup
+uv tool install "pipecat-ai[cli]" --with pipecat-ai-context-hub
+pipecat mcp install
```
Comment (kompfner) [inline src/pipecat/cli/agent_templates/GETTING_STARTED.md]: If we're calling the product "context hub", isn't the fact that it's an MCP server just an implementation detail?

Idea (from @markbackman):
- `pipecat context-hub install|refresh`
- `pc ch install|refresh` (all shorthands)

**Comment by kompfner** [comment]
A few questions/comments, just based on reviewing the PR description:

- Do we ever output something to the user telling them "hey, please restart your coding agent for the context hub to become available"? I didn't see it in the PR description's sample CLI output.
- What happens if you answer "no" to "Build it now?" Does it tell you what command to run to built the index later?
- I feel like the line "Your agent reads current Pipecat APIs from a local index, not training data" could be tweaked a bit to include some "why" (local index will give your agent more accurate, up-to-date info)

**Comment by markbackman** [comment]
> * Do we ever output something to the user telling them "hey, please restart your coding agent for the context hub to become available"? I didn't see it in the PR description's sample CLI output.

What case do you have in mind? The user is running `pipecat init`, which attaches the MCP tools to the coding agent, so no restart is needed. Are you thinking if you have two terminals--one with a coding agent and one running pipecat init--that you need to restart one?

Maybe a line in GETTING_STARTED.md is the right place for this guidance? WDYT?

> * What happens if you answer "no" to "Build it now?" Does it tell you what command to run to built the index later?

Yes, it does tell you if you say "no" with the following:
```
Build the index before your first coding session:
    pipecat context-hub refresh
```

> * I feel like the line "Your agent reads current Pipecat APIs from a local index, not training data" could be tweaked a bit to include some "why" (local index will give your agent more accurate, up-to-date info)

Yeah, totally. I tried to polish this line for about 15 minutes yesterday and never landed on something I love. Any suggestions?

If we want some "why", we could try:
```
Your agent reads a local index of current Pipecat docs and API source,
so it writes working code instead of guessing from training data.
```

I was trying to keep it to a single line. Any options to suggest?

**Comment by kompfner** [comment]
> The user is running pipecat init, which attaches the MCP tools to the coding agent, so no restart is needed.

I was thinking of the scenario where you already had a coding agent open and *it* was running the `pc init` on your behalf.

> Yes, it does tell you if you say "no" with the following

👍 

> Any suggestions?

```
Your agent uses up-to-date knowledge of Pipecat APIs from Context Hub's local index, rather relying on stale training data.
```

I also don't mind the one you had, above. In my suggestion I use "Context Hub" instead of just "a local index" to draw the connection for the reader between the index and the previous line ("Registered the Context Hub MCP server"); feels like it could be helpful for building a mental model.

**Comment by markbackman** [comment]
> I was thinking of the scenario where you already had a coding agent open and it was running the pc init on your behalf.

Ok, finally coming back to this. Since this is the agent running `pipecat init` for you, I'm opting to update the AGENTS.md with an instruction for the agent. It's told that the hub's CLI is available, but to ask the user if they want to restart the session to set up the hub database for MCP commands:
https://github.com/pipecat-ai/pipecat/pull/5122/changes/c5227313a28912484a4cb332c4490158a0fc95aa#diff-576ba06fa7e9713d9823f913c9cb9b73b714ed5ceff07432a2dfd320dfe4b140R48-R57

> Your agent uses up-to-date knowledge of Pipecat APIs from Context Hub's local index, rather relying on stale training data.

Nice, this is clear. I'll use this.

**Comment by markbackman** [comment]
@kompfner This PR should be ready for review. I'm waiting on a Context Hub release from Varun to land before we can ship this.

**Comment by vr000m** [comment]
GH actions are still flaky. However I manually shipped 0.5.1. So everything should be available now.

**Comment by vr000m** [comment]
Another thought while looking at the code. 

Since PCH is installed in global (~/.local/pipecat-contrxt-hub). Are the addon repos set in the shell environment or should PCH consider a config yaml that Pipecat cli can set to add custom Pipecat repos (like the customer's code which may have custom processors, etc.).

**Comment by markbackman** [comment]
> Another thought while looking at the code.
> 
> Since PCH is installed in global (~/.local/pipecat-contrxt-hub). Are the addon repos set in the shell environment or should PCH consider a config yaml that Pipecat cli can set to add custom Pipecat repos (like the customer's code which may have custom processors, etc.).

This configuration info is getting overridden. I think a config yaml is probably the right solution. Since the solution is solely in Context Hub, I'll have Claude draft thoughts for you and then you can work on this in Pipecat Context Hub.

It doesn't block this path moving forward. We can just bump the minimum version once you've shipped an improvement.

Claude comments incoming...

**Comment by markbackman** [comment]
@vr000m Good catch — and it's sharper than a config-ergonomics question. `_load_dotenv()` reads `.env` from the **current directory**, while the index is machine-global, and `refresh` deletes any repo not in that invocation's `effective_repos`:

```python
configured = set(config.sources.effective_repos)
for meta_key in all_meta:
    if slug not in configured:
        logger.info("Repo %s no longer configured, cleaning up", slug)
        await index_store.delete_by_repo(slug)
```

So with a `.env` in a customer project:

1. `cd customer-a && refresh` → their code is indexed
2. `cd anywhere-else && refresh` → **deleted**, because that directory's environment doesn't name it
3. `cd customer-a && refresh` → re-cloned and re-embedded

The global index ends up tracking whichever directory you last refreshed from.

**Root cause is a scope mismatch, not a bad prune.** That cleanup is correct *given its assumption* — that `effective_repos` is the complete intended set — and the assumption only holds if the config is as global as the index. Right now the index is per-machine and the config is per-directory.

**To get the obvious question out of the way: `pipecat init` is not doing this to anyone.** Pipecat invokes the hub exactly twice, and neither can prune:

- `install --no-refresh` — no refresh runs at all
- the index build — gated behind `if metadata.get("last_refresh_at"): return`, so it only fires when no index exists, and there's nothing to delete

Where this PR *is* adjacent: the new freshness notice tells you to run `pipecat context-hub refresh`, which you'll do from wherever you're standing — that's the trigger — and bundling the hub into `pipecat-ai[cli]` puts it on many more machines. So the PR widens the audience and supplies the prompt without ever performing the destructive operation. It isn't blocked by this (all the code is hub-side, and we pin `>=0.5.1,<1`, so a 0.5.2 reaches users with no pipecat release), but it does argue for fixing it sooner rather than later.

Options as I see them:

1. **Stop deleting silently.** Warn — "repo X is indexed but not configured here; `refresh --prune` to remove" — and only delete when asked. Smallest change, removes the data loss, designs nothing.
2. **Remember the configured set** in index metadata and union it with the current environment. Non-destructive by default, but accumulates with no way to inspect or edit the set — which really argues for 3.
3. **A global config file**, as you suggested — repo list under `~/.pipecat-context-hub/`, so config scope matches index scope. The env var becomes a one-off override rather than the system of record, and it can grow an `add-repo` command that surfaces as `pipecat context-hub add-repo` for free through the existing plugin entry point.
4. **Per-project indexes** via `PIPECAT_HUB_DATA_DIR`. Already works, and the only option giving genuine isolation — at ~900 MB apiece. Probably a documented escape hatch rather than a default.

My inclination is 1 as a quick 0.5.2 safety fix and 3 as the real answer; they compose, since 3 makes 1's warning rare.

One thing worth deciding explicitly either way: with a single global index, every configured repo is searchable from every project. Fine for most people, wrong for anyone working across clients — and today that's an accident of the design rather than a decision. Do we want to state it as intended behaviour, or is 4 the sanctioned answer for that case?

**Comment by vr000m** [comment]
I think the code that you're writing is managed by Claude already so context hub unless refreshed will lag. 

And if cross pollination of different projects is bad, I think excluding the repo is the right choice. 

My thesis is that we have one global config which can add on to the hub defaults

**Comment by markbackman** [comment]
> My thesis is that we have one global config which can add on to the hub defaults

This sounds right!

**Comment by vr000m** [comment]
https://github.com/pipecat-ai/pipecat-context-hub/pull/114#issuecomment-5230568683, also releasing `0.5.2`

**Comment by kompfner** [comment]
Still interested in:

> the scenario where you already had a coding agent open and it was running the pc init on your behalf

So, like, instead of explicitly running `pc init` to bootstrap a project, you just open up Claude Code and say "Look up how to initialize a Pipecat project, then do that for me." If your coding agent then runs `pc init` on your behalf, it'll then hit a point where it needs to tell you "hey, you'll need to restart me in order to give me access to the Pipecat Context Hub". That's a flow I'm interested in seeing work.

Other than that, the changes here look good! I'll go ahead and pre-approve this PR, pending this last scenario being handled.

**Review by kompfner — APPROVED** [review]
